### PR TITLE
feat(server): dev safety checks

### LINKED ISSUE #1416: Introduce development time checks to remove remaining foot guns in the project
I read our documentation and found the following assumptions in the getting started and concepts pages. But we don't check for any of these to give good error messages.

- Only one _app.tsx is allowed per application and it must have default export
- One 500.tsx and it must have default export
- One 404.tsx and it must have default export
- Pascal case or kebab case islands (island name)
- Route must have handler or component
- If static override, ensure there is no normal static directory
- Static file conflict with routes (because static files have higher priority)
- Plugin doesn't call render
- Plugin Each defined module must be a JavaScript/TypeScript module

As an example, if you remove the default export from `_app.tsx` then the following unhelpful error message is presented (note how it doesn't mention `_app.tsx`:
```
An error occurred during route handling or page rendering. Error: Undefined component passed to createElement()

You likely forgot to export your component or might have mixed up default and named imports<#text params="[object Object]" url="http://localhost:8000/" route="/" data="undefined" Component="function Component() {}" />

  in #text

    at i.__b (https://esm.sh/stable/preact@10.15.1/denonext/debug.js:11:362)
    at k (https://esm.sh/v125/preact-render-to-string@6.1.0/X-ZS8q/denonext/preact-render-to-string.mjs:2:2086)
    at k (https://esm.sh/v125/preact-render-to-string@6.1.0/X-ZS8q/denonext/preact-render-to-string.mjs:2:2591)
    at k (https://esm.sh/v125/preact-render-to-string@6.1.0/X-ZS8q/denonext/preact-render-to-string.mjs:2:2591)
    at T (https://esm.sh/v125/preact-render-to-string@6.1.0/X-ZS8q/denonext/preact-render-to-string.mjs:2:1312)
    at renderInner (https://deno.land/x/fresh@1.2.0/src/server/render.ts:191:16)
    at renderSync (https://deno.land/x/fresh@1.2.0/src/server/render.ts:209:7)
    at https://deno.land/x/fresh@1.2.0/src/server/render.ts:240:38
    at Object.DEFAULT_RENDER_FN [as renderFn] (https://deno.land/x/fresh@1.2.0/src/server/context.ts:792:3)
    at renderAsync (https://deno.land/x/fresh@1.2.0/src/server/render.ts:240:18)
Error: Undefined component passed to createElement()

You likely forgot to export your component or might have mixed up default and named imports<#text params="[object Object]" url="http://localhost:8000/" route="" data="undefined" Component="function Component() {}" />

  in #text
  in Provider

    at i.__b (https://esm.sh/stable/preact@10.15.1/denonext/debug.js:11:362)
    at k (https://esm.sh/v125/preact-render-to-string@6.1.0/X-ZS8q/denonext/preact-render-to-string.mjs:2:2086)
    at k (https://esm.sh/v125/preact-render-to-string@6.1.0/X-ZS8q/denonext/preact-render-to-string.mjs:2:2591)
    at k (https://esm.sh/v125/preact-render-to-string@6.1.0/X-ZS8q/denonext/preact-render-to-string.mjs:2:2591)
    at T (https://esm.sh/v125/preact-render-to-string@6.1.0/X-ZS8q/denonext/preact-render-to-string.mjs:2:1312)
    at renderInner (https://deno.land/x/fresh@1.2.0/src/server/render.ts:191:16)
    at renderSync (https://deno.land/x/fresh@1.2.0/src/server/render.ts:209:7)
```

We should include some sort of development time set of checks to look for these sorts of misconfigurations / misunderstandings and let the developer know, so they don't spend time having to debug.

### LINKED ISSUE #1524: Conflicts in the Routing Process.
Currently 
It seems as if `/users/index.tsx` and `users.tsx` are in conflict.
Are there any official operational techniques...?

### PR DESCRIPTION
**Pull request by mbhrznr**
closes #1416  
closes #1524

as discussed in the aforementioned issue, @deer was so kind to let me finish this one off.
as expected it gave me quite some nice understanding of `fresh`'s internals.

this pr introduces the following check when being in `dev` mode:  
- [x] only one _app.tsx is allowed per application and it must have default export
- [x] one 500.tsx and it must have default export
- [x] one 404.tsx and it must have default export
- [x] ~~pascal case or kebab case islands (island name)~~ (this one became obsolete w/ 1.3)
- [x] route must have handler or component
- [x] route patterns must be unique 
- [x] if static override, ensure there is no normal static directory
- [x] static file conflict with routes (because static files have higher priority)
- [x] plugin doesn't call render
- [x] plugin each defined module must be a js/ts module

the functionality can be tested by navigating into `tests/fixture_dev_checks/` and running `deno task start`

### CODE REVIEW

File: src/server/context.ts [hunk 1]
```diff
@@ -359,6 +362,30 @@ export class ServerContext {
...
+        () => assertModuleExportsDefault(app, "_app"),
+        () => assertSingleModule(routes, "_app"),
+        () => assertModuleExportsDefault(unknownModule, "_404"),
+        () => assertSingleModule(routes, "_404"),
+        () => assertModuleExportsDefault(errorModule, "_500"),
+        () => assertSingleModule(routes, "_500"),
+        () => assertRoutesHaveHandlerOrComponent(routes),
+        () => assertStaticDirSafety(opts.staticDir ?? "", defaultStaticDir),
+        () => assertNoStaticRouteConflicts(routes, staticFiles),
+        () => assertPluginsCallRender(opts.plugins ?? []),
```
Comment (mbhrznr) [inline src/server/context.ts:394]: apparently these plugin checks currently cause the following issue in the ci:
```sh
error: Leaking async ops:
  - 1 async operation to sleep for a duration was started before this test, but was completed during the test. Async operations should not complete in a test if they were not started in that test.
            This is often caused by not cancelling a `setTimeout` or `setInterval` call.
```

would it make sense to add yet another check for testing in `isDevMode`?

**Comment by deer** [comment]
Thanks for finishing this up. Looks pretty good, but I wonder if we want to specifically create a union type for all the categories and then reference them as something like this:
```tsx
return [{
      category: DevChecks.ModuleExport,
      message: `Your ${moduleName} file does not have a default export.`,
      fileLink: moduleName,
    }];
```
instead of
```tsx
return [{
      category: "Module Export",
      message: `Your ${moduleName} file does not have a default export.`,
      fileLink: moduleName,
    }];
```

Additionally, what about implementing the check in https://github.com/denoland/fresh/pull/1414? And taking the tests as well. Then I can just close that one, and a maintainer can close https://github.com/denoland/fresh/pull/539. And then you can resolve three issues at once (the two you mention, plus 475 which I was going to resolve).

**Comment by mbhrznr** [comment]
@deer thanks for the suggestion. will incorporate everything you mentioned!

**Review by marvinhagemeister — APPROVED** [review]
LGTM, thanks for adding these 🎉

**Comment by marvinhagemeister** [comment]
Apologies for the churn. I merged the PR too quickly and it's currently blocking me on feature work, so I've reverted it in #1567. So far I encountered that the handlers rule doesn't account for them being an array.

The output is a bit messy when there are multiple errors. Maybe some visual hierarchy via colors can help?

Most rules don't print the file name where the error happened which makes it difficult to track down where the error is coming from. The `route.name` property that many rules rely on is not the file name.

**Pull request merged**